tests/load/test_freeze_and_data_contract.py

```diff
@@ -0,0 +1,591 @@
+        "data_type": "discard_row"
+    }})
+    table_counts = load_table_counts(pipeline, *[t["name"] for t in pipeline.default_schema.data_tables()])
+    assert table_counts["items"] == 10
+
+    # now allow
+    run_resource(pipeline, items_with_variant, {setting_location: {
+        "data_type": "evolve"
+    }})
+    table_counts = load_table_counts(pipeline, *[t["name"] for t in pipeline.default_schema.data_tables()])
+    assert table_counts["items"] == 20
+
+@pytest.mark.parametrize("setting_location", LOCATIONS)
+def test_single_settings_value(setting_location: str) -> None:
+    pipeline = get_pipeline()
+
+    run_resource(pipeline, items, {})
+    table_counts = load_table_counts(pipeline, *[t["name"] for t in pipeline.default_schema.data_tables()])
+    assert table_counts["items"] == 10
+
+    # trying to add variant when forbidden will fail
+    run_resource(pipeline, items_with_variant, {setting_location: "discard_row"})
+    table_counts = load_table_counts(pipeline, *[t["name"] for t in pipeline.default_schema.data_tables()])
+    assert table_counts["items"] == 10
+
+    # trying to add new column will fail
+    run_resource(pipeline, items_with_new_column, {setting_location: "discard_row"})
+    table_counts = load_table_counts(pipeline, *[t["name"] for t in pipeline.default_schema.data_tables()])
+    assert table_counts["items"] == 10
+
+    # trying to add new table will fail
+    run_resource(pipeline, new_items, {setting_location: "discard_row"})
+    table_counts = load_table_counts(pipeline, *[t["name"] for t in pipeline.default_schema.data_tables()])
+    assert table_counts["items"] == 10
+    assert "new_items" not in table_counts
+
+
+def test_data_contract_interaction() -> None:
+    """
+    ensure data contracts with pydantic are enforced properly
+    """
+    from pydantic import BaseModel
+
+    class Items(BaseModel):
+        id: int  # noqa: A003
+        name: Optional[str]
+        amount: Optional[int]
+
+    @dlt.resource(name="items", columns=Items)
+    def get_items_simple():
+        yield from [{
+            "id": 5
+        }]
+
+    @dlt.resource(name="items", columns=Items)
+    def get_items():
+        yield from [{
+            "id": 5,
+            "name": "dave",
+        }]
+
+    @dlt.resource(name="items")
+    def get_items_variant():
+        yield from [{
+            "id": 5,
+            "name": "dave",
+            "amount": "HELLO"
+        }]
+
+    @dlt.resource(name="items")
+    def get_items_new_col():
+        yield from [{
+            "id": 5,
+            "name": "dave",
+            "amount": 6,
+            "new_col": "hello"
+        }]
+
+    @dlt.resource(name="items")
+    def get_items_subtable():
+        yield from [{
+            "id": 5,
+            "name": "dave",
+            "amount": 6,
+            "sub": [{"hello": "dave"}]
+        }]
+
+    # test variants
+    pipeline = get_pipeline()
+    pipeline.run([get_items_simple()])
```

Comment:
@rudolfix for the tests to work and the populated attribute to be set we always have to do one load. should we maybe rename the populated hint to something else (but with the same functionality) and set it to true in the schema if a complete table schema is supplied via columns? I think that would be more intuitive. Of course we can also do apply_hints, but this seems a bit intuitive for the enduser.